### PR TITLE
semLambda removed, semProcAux reworked

### PR DESCRIPTION
**Pull request by saem**
Changes:
* removed redundancy in semProcAux
* lambdas are also checked for triggering compile time
* calling conventions are now more strictly set
* nullary and non-generic procs are differentiated correctly

Remarks on code style:
The code style is not the same, of course. The intention here is that control flow
does not act to combine derived state. This way judgements and decisions are
disentangled. The use of `let` over say `template` which evaluate at each invocation
is to ideally bring the point of failure closer to the earliest point possible where the
stale judgement was made. This also assists in debugging as such intermediate state
is excellent for inspection either via debug or quick echos.

I'll squash the commits before hand, as the commit messages aren't pretty.

PS. Making the body handling more linear like the rest of the logic is for later PRs.

Many thanks to @Clyybber we paired intensely on this and it's a shame that github
isn't recognizing the commiters correctly.

### CODE REVIEW

**Comment by saem** [comment]
I'll wait for the first pass of CI to see what broke, fix and rebase as needed.

**Comment by saem** [comment]
Looks like CI is having issues unrelated to this PR.

Should be good to review regardless.

**Comment by saem** [comment]
@Clyybber let me know if this feels more consistent with the `isGenericParams`. I inlined them as those should be quick queries, also figured the nil checks are so cheap considering they preceed a load there is basically no benefit to not having them.

The `isAnon` and `isLambda` seem like they're very close, the only thing to figure out is why `isLambda` needs guards where `isAnon` was working fine before.

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -1908,47 +1867,66 @@ proc semProcAux(c: PContext, n: PNode, kind: TSymKind,
...
-    elif kind in OverloadableSyms:
-      if not typeIsDetermined:
-        addInterfaceOverloadableSymAt(c, oldScope, s)
-    else:
-      if not typeIsDetermined:
-        addInterfaceDeclAt(c, oldScope, s)
-    if n[pragmasPos].kind != nkEmpty:
-      pragma(c, s, n[pragmasPos], validPragmas)
+
+  if not hasProto and sfGenSym notin s.flags: #and not isAnon:
```
Comment (saem) [inline compiler/semstmts.nim:1898]: Can we have a gensymmed name that was used multiple times for emitting overloaded definitions?

Comment (Clyybber) [reply compiler/semstmts.nim:1898]: At this point each overload is a seperate symbol

**Comment by saem** [comment]
Nevermind, previous message was all wrong, the global variable reference screwed everything up.

**Comment by saem** [comment]
@Clyybber I think this is about as done as it's going to be, either that or I'm really tired and can't see things any more.

Do we need another reviewer and we're good to go you think?

**Pull request merged**